test/e2e/storage/in_tree_volumes.go

```diff
@@ -0,0 +1,455 @@
+	}
+
+	// Check if fsType is supporetd by driver
+	if _, ok := driverInfo.supportedFsType[testPattern.fsType]; !ok {
+		framework.Skipf("Driver %s doesn't support %v -- skipping", driverInfo.name, testPattern.fsType)
+	}
+
+	// Check with driver specific logic
+	driver.skipUnsupportedTest(testPattern)
+
+	//  Check with testSuite specific logic
+	testSuite.skipUnsupportedTest(testPattern, driver)
+}
+
+// Generic volume test resource
+type genericVolumeTestResource struct {
+	driver    testDriver
+	volType   string
+	volSource *v1.VolumeSource
+	pvc       *v1.PersistentVolumeClaim
+	pv        *v1.PersistentVolume
+	sc        *storagev1.StorageClass
+}
+
+func createVolumeSourceWithPVCPV(
+	f *framework.Framework,
+	name string,
+	pvSource *v1.PersistentVolumeSource,
+	readOnly bool,
+) (*v1.VolumeSource, *v1.PersistentVolume, *v1.PersistentVolumeClaim) {
+	pvConfig := framework.PersistentVolumeConfig{
+		NamePrefix:       fmt.Sprintf("%s-", name),
+		StorageClassName: f.Namespace.Name,
+		PVSource:         *pvSource,
+	}
+	pvcConfig := framework.PersistentVolumeClaimConfig{
+		StorageClassName: &f.Namespace.Name,
+	}
+
+	framework.Logf("Creating PVC and PV")
+	pv, pvc, err := framework.CreatePVCPV(f.ClientSet, pvConfig, pvcConfig, f.Namespace.Name, false)
+	Expect(err).NotTo(HaveOccurred(), "PVC, PV creation failed")
+
+	err = framework.WaitOnPVandPVC(f.ClientSet, f.Namespace.Name, pv, pvc)
+	Expect(err).NotTo(HaveOccurred(), "PVC, PV failed to bind")
+
+	volSource := &v1.VolumeSource{
+		PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+			ClaimName: pvc.Name,
+			ReadOnly:  readOnly,
+		},
+	}
+	return volSource, pv, pvc
+}
+
+func createVolumeSourceWithPVCPVFromDynamicProvisionSC(
+	f *framework.Framework,
+	name string,
+	claimSize string,
+	sc *storagev1.StorageClass,
+	readOnly bool,
+	volMode *v1.PersistentVolumeMode,
+) (*v1.VolumeSource, *v1.PersistentVolume, *v1.PersistentVolumeClaim) {
+	cs := f.ClientSet
+	ns := f.Namespace.Name
+
+	By("creating a claim")
+	pvc := getClaim(claimSize, ns)
+	pvc.Spec.StorageClassName = &sc.Name
+	if volMode != nil {
+		pvc.Spec.VolumeMode = volMode
+	}
+
+	var err error
+	pvc, err = cs.CoreV1().PersistentVolumeClaims(ns).Create(pvc)
+	Expect(err).NotTo(HaveOccurred())
+
+	err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, cs, pvc.Namespace, pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
+	Expect(err).NotTo(HaveOccurred())
+
+	pvc, err = cs.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	pv, err := cs.CoreV1().PersistentVolumes().Get(pvc.Spec.VolumeName, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	volSource := &v1.VolumeSource{
+		PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+			ClaimName: pvc.Name,
+			ReadOnly:  readOnly,
+		},
+	}
+	return volSource, pv, pvc
+}
+
+func (r *genericVolumeTestResource) setup(driver testDriver, testPattern testPattern) {
```

Comment:
nit: can we use a more descriptive name than "setup". what is this setting up? the driver, the test, the framework?

Reply:
Renamed as below:

```Setup``` -> ```SetupResource```
```Cleanup``` ->   ```CleanupResource```